File: examples/multiple_simple_compression.c
/*
 * Copyright (c) 2016-present, Yann Collet, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under both the BSD-style license (found in the
 * LICENSE file in the root directory of this source tree) and the GPLv2 (found
 * in the COPYING file in the root directory of this source tree).
 * You may select, at your option, one of the above-listed licenses.
 */

#include <stdlib.h>    // malloc, free, exit
#include <stdio.h>     // fprintf, perror, fopen, etc.
#include <string.h>    // strlen, strcat, memset, strerror
#include <errno.h>     // errno
#include <sys/stat.h>  // stat
#include <zstd.h>      // presumes zstd library is installed
#include "utils.h"

typedef struct {
    void* fBuffer;
    void* cBuffer;
    size_t fBufferSize;
    size_t cBufferSize;
    ZSTD_CCtx* cctx;
} resources;

/*
 * allocate memory for buffers big enough to compress all files
 * as well as memory for output file name (ofn)
 */
static resources createResources_orDie(int argc, const char** argv, char **ofn, int* ofnBufferLen)
{
    size_t maxFilenameLength=0;
    size_t maxFileSize = 0;

    int argNb;
    for (argNb = 1; argNb < argc; argNb++) {
      const char* const filename = argv[argNb];
      size_t const filenameLength = strlen(filename);
      size_t const fileSize = fsize_orDie(filename);

      if (filenameLength > maxFilenameLength) maxFilenameLength = filenameLength;
      if (fileSize > maxFileSize) maxFileSize = fileSize;
    }

    resources ress;
    ress.fBufferSize = maxFileSize;
    ress.cBufferSize = ZSTD_compressBound(maxFileSize);

    *ofnBufferLen = maxFilenameLength + 5;
    *ofn = (char*)malloc_orDie(*ofnBufferLen);
    ress.fBuffer = malloc_orDie(ress.fBufferSize);
    ress.cBuffer = malloc_orDie(ress.cBufferSize);
    ress.cctx = ZSTD_createCCtx();
    if (ress.cctx==NULL) { fprintf(stderr, "ZSTD_createCCtx() error \n"); exit(10); }
    return ress;
}

static void freeResources(resources ress, char *outFilename)
{
    free(ress.fBuffer);
    free(ress.cBuffer);
    ZSTD_freeCCtx(ress.cctx);   /* never fails */
    free(outFilename);
}

/* compress with pre-allocated context (ZSTD_CCtx) and input/output buffers*/
static void compressFile_orDie(resources ress, const char* fname, const char* oname)
{
    size_t fSize = loadFile_orDie(fname, ress.fBuffer, ress.fBufferSize);

    size_t const cSize = ZSTD_compressCCtx(ress.cctx, ress.cBuffer, ress.cBufferSize, ress.fBuffer, fSize, 1);
    if (ZSTD_isError(cSize)) {
        fprintf(stderr, "error compressing %s : %s \n", fname, ZSTD_getErrorName(cSize));
        exit(8);
    }

    saveFile_orDie(oname, ress.cBuffer, cSize);

    /* success */
    // printf("%25s : %6u -> %7u - %s \n", fname, (unsigned)fSize, (unsigned)cSize, oname);
}

int main(int argc, const char** argv)
{
    const char* const exeName = argv[0];

    if (argc<2) {
        printf("wrong arguments\n");
        printf("usage:\n");
        printf("%s FILE(s)\n", exeName);
        return 1;
    }

    /* memory allocation for outFilename and resources */
    char* outFilename;
    int outFilenameBufferLen;
    resources const ress = createResources_orDie(argc, argv, &outFilename, &outFilenameBufferLen); 

    /* compress files with shared context, input and output buffers */
    int argNb;
    for (argNb = 1; argNb < argc; argNb++) {
        const char* const inFilename = argv[argNb];
        int inFilenameLen = strlen(inFilename);
        assert(inFilenameLen + 5 <= outFilenameBufferLen);
        memcpy(outFilename, inFilename, inFilenameLen);
        memcpy(outFilename+inFilenameLen, ".zst", 5);
        compressFile_orDie(ress, inFilename, outFilename);
    }

    /* free momery */
    freeResources(ress,outFilename);

    printf("compressed %i files \n", argc-1);

    return 0;
}

File: examples/multiple_simple_compression.zig
const std = @import("std");
const zstd = @import("zstd");
const common = @import("common.zig");

pub fn main() !void {
    const allocator = std.heap.c_allocator;
    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    if (args.len < 2) {
        std.debug.print("wrong arguments\nusage:\n{s} FILE(s)\n", .{args[0]});
        return;
    }

    // Pre-calculate max file size and max filename length
    var max_file_size: usize = 0;

    var i: usize = 1;
    while (i < args.len) : (i += 1) {
        const file = try std.fs.cwd().openFile(args[i], .{});
        const stat = try file.stat();
        file.close();

        if (stat.size > max_file_size) max_file_size = @intCast(stat.size);
    }

    const c_buffer_size = zstd.c.ZSTD_compressBound(max_file_size);
    const f_buffer = try allocator.alloc(u8, max_file_size);
    defer allocator.free(f_buffer);

    const c_buffer = try allocator.alloc(u8, c_buffer_size);
    defer allocator.free(c_buffer);

    const cctx = zstd.c.ZSTD_createCCtx();
    if (cctx == null) {
        std.debug.print("zstd.eateCCtx() failed!\n", .{});
        return;
    }
    defer _ = zstd.c.ZSTD_freeCCtx(cctx);

    i = 1;
    while (i < args.len) : (i += 1) {
        const input_filename = args[i];

        // Load file into shared buffer
        const file = try std.fs.cwd().openFile(input_filename, .{});
        const f_size = try file.readAll(f_buffer);
        file.close();

        // Compress
        const cSize = zstd.c.ZSTD_compressCCtx(cctx, c_buffer.ptr, c_buffer_size, f_buffer.ptr, f_size, 1);
        if (zstd.c.ZSTD_isError(cSize) != 0) {
            std.debug.print("error compressing: {s}\n", .{zstd.c.ZSTD_getErrorName(cSize)});
            continue;
        }

        const out_filename = try common.createOutFilename(allocator, input_filename);
        defer allocator.free(out_filename);
        try common.writeFile(out_filename, c_buffer[0..cSize]);

        std.debug.print("{s} : {d} -> {d} - {s} \n", .{
            input_filename,
            f_size,
            cSize,
            out_filename,
        });
    }

    std.debug.print("compressed {d} files \n", .{args.len - 1});
}



